### PR TITLE
Fix issue 16486: Alias template function parameter resolution

### PR DESCRIPTION
**Pull request by baziotis**
It probably needs more testcases.

### CODE REVIEW

File: src/dmd/dtemplate.d [hunk 1]
```diff
@@ -1351,8 +1351,37 @@ extern (C++) final class TemplateDeclaration : ScopeDsymbol
...
+                Type prmtype = fparam.type;
+                //printf("prmtype: %s\n", prmtype.toChars());
+                while(prmtype.ty == Tinstance)
+                {
+                    TypeInstance prmti = cast(TypeInstance) prmtype;
+                    //printf("prmti: %s\n", prmti.toChars());
+                    if(prmti.tempinst.findTempDecl(sc, null))
+                    {
+                        TemplateDeclaration td = prmti.tempinst.tempdecl.isTemplateDeclaration();
+                        if(td)
```
Comment (Basile-z) [inline src/dmd/dtemplate.d]: You can put the declaration directly as condition `if( TemplateDeclaration td = prtmi....)`

File: src/dmd/dtemplate.d [hunk 2]
```diff
@@ -1351,8 +1351,37 @@ extern (C++) final class TemplateDeclaration : ScopeDsymbol
...
+                        {
+                            //printf("td: %s\n", td.toChars());
+                            if(td.onemember)
+                            {
+                                AliasDeclaration ad = td.onemember.isAliasDeclaration();
+                                if(ad)
```
Comment (Basile-z) [inline src/dmd/dtemplate.d]: same here, `if (AliasDeclaration ad = td.onemember.isAliasDeclaration())`

**Review by ghost — COMMENTED** [review]
Just two advice on the style . Also take care because I see that you made your changes in the master branch of your fork. It's usually much more convenient to create a branch for each PR you want to propose. Using master will lead to maintenance problems: hard resets, force pushes, impossibility to have two PRs at the same time, etc.

File: .gitignore [hunk 3]
```diff
@@ -17,6 +17,7 @@ generated/
 *.pdb
 src/vcbuild/Win32
 src/vcbuild/x64
+src/vcbuild/.vs
```
Comment (wilzbach) [inline .gitignore]: Please do this in a separate PR. It's unrelated.

File: src/dmd/dtemplate.d [hunk 4]
```diff
@@ -1351,8 +1351,37 @@ extern (C++) final class TemplateDeclaration : ScopeDsymbol
             {
                 Parameter fparam = fparameters[parami];
 
+                Type prmtype = fparam.type;
+                //printf("prmtype: %s\n", prmtype.toChars());
+                while(prmtype.ty == Tinstance)
+                {
```
Comment (wilzbach) [inline src/dmd/dtemplate.d]: Space after keywords like while or if - checkout the DStyle at dlang.org/dstyle

File: src/dmd/dtemplate.d (hunk 2 above)
Comment (baziotis) [reply src/dmd/dtemplate.d]: Got it, for both, I just thought it would be too long for a condition. Also yes, bad decision to go with the master.

File: test/compilable/test16486.d [hunk 5]
```diff
@@ -0,0 +1,35 @@
+// Example 1
```
Comment (wilzbach) [inline test/compilable/test16486.d]: Maybe use more descriptive names than example 1,2, 3?

**Comment by baziotis** [comment]
All the tests seem to fail due to timeout. Only now I realized how easily this loop can run forever the way it's written. I will fix it tomorrow along with the other proposals.
